tests/scenario/test_charm.py

```diff
@@ -0,0 +1,601 @@
+        },
+    )
+    storage = scenario.Storage("local-media")
+    peer_relation = scenario.PeerRelation(
+        "takahe-peer", "takahe-peer", local_app_data={"secret-id": secret.id}
+    )
+    database_relation = scenario.Relation(
+        "database", "postgresql_client", remote_app_data={"endpoints": endpoint}
+    )
+    # SCENARIO-NOTE: It would be nice if the charm test didn't need to know the
+    # layout, since the lib is hiding all of that - but maybe that's really an
+    # issue with how the lib hides everything.
+    db_secret = scenario.Secret(
+        "1235",
+        {0: {"username": user, "password": password}},
+        label=f"database.{database_relation.relation_id}.user.secret",
+    )
+    state = scenario.State(
+        leader=True,
+        relations=[peer_relation, database_relation],
+        containers=[web_container, background_container],
+        storage=[storage],
+        secrets=[secret, db_secret],
+        config={"main-domain": domain, "media-uri": media},
+    )
+
+    ## Act.
+    out = ctx.run("upgrade_charm", state)
+
+    ## Assert.
+    cmd = (
+        "gunicorn takahe.wsgi:application -b 0.0.0.0:8001 "
+        "--access-logfile access.log --error-logfile error.log"
+    )
+    assert_running_state(
+        ctx,
+        out,
+        out.containers[0],
+        user,
+        password,
+        endpoint,
+        secret.contents[0]["takahe-secret-key"],
+        domain,
+        media,
+        "Takahē web server",
+        cmd,
+        web_container.name,
+    )
+    assert_running_state(
+        ctx,
+        out,
+        out.containers[1],
+        user,
+        password,
+        endpoint,
+        secret.contents[0]["takahe-secret-key"],
+        domain,
+        media,
+        "Takahē background service",
+        "python3 manage.py runstator",
+        background_container.name,
+    )
+    assert out.workload_version == version
+    # SCENARIO-NOTE: it feels a bit odd to have to include the UnknownStatus
+    # here, although I understand that it's because it's what was set (or not
+    # in this case) in the input state).)
+    assert ctx.unit_status_history == [
+        ops.UnknownStatus(),
+        ops.MaintenanceStatus("Upgrading database tables..."),
+        ops.MaintenanceStatus("Restarting service..."),
+    ]
+
+
+def test_database_ready(ctx):
+    ## Arrange.
+    version = "1.2.3"
+    domain = "aotearoa.dev"
+    media = "local://"
+    user = "USER"
+    password = "PASS"
+    endpoint = "db.local:8000"
+    secret = scenario.Secret("1234", {0: {"takahe-secret-key": "friend"}})
+    web_container = scenario.Container(
+        "takahe-web",
+        can_connect=True,
+    )
+    background_container = scenario.Container(
+        "takahe-background",
+        can_connect=True,
+        exec_mock={
+            ("python", "-c", "import takahe;print(takahe.__version__)"): scenario.ExecOutput(
```

Comment:
I find the `exec_mock` somewhat awkward, but maybe it's inevitable. Does it do prefix-matching like `Harness.handle_exec`, or only full commands? Anything, worth considering if we can improve this in Scenario.

Reply:
> Does it do prefix-matching like `Harness.handle_exec`, or only full commands?

Full command - it has to be an exact match. Basically, it's doing a `get` in the provided dictionary.

> Anything, worth considering if we can improve this in Scenario.

I do feel like exec and files are an area where Harness does have an edge over Scenario. Would be good to talk about this in our Scenario chat (I've made a note in that doc).